File: srcs/t1.f
 
 
 : for ['] >r , begin ; immediate
 : next ['] r> , ['] lit , [ 1 ] , ['] - , 
   ['] dup , ['] 0# , ['] ?branch ,
   here - , ['] drop , ; immediate

 : hook begin >r ;
 : back r> again ;
 : ?back r> until ;

 : swip >r swap r> ; 
 : rot swip swap ;
 : -rot swap swip ;
 : flip swap swip swap ;

 : dovar r> dup cell + >r ;
 : docon r> dup cell + >r @ ;
 : literal lit lit , , ;
 : execute >r ;

 : sp0 lit [ sp@ , ] ;

 : rp0 lit [ rp@ , ] ;

 : && sp@ . drop rp@ . drop latest @ . drop heap @ . drop ;
 
 : :name here : 0 state ! ;
 
 : :noname here 1 state ! ;

 : create here : 
   ['] lit , here cell cell + + , ['] exit , 
   0 state ! latest ! ;

 : does ;

 : does> ;

 : <builds create 0 , ;

 : variable create cell allot ;

 : array create allot ;

 : hash :name dup heap ! cell + @ ;
 
 : seek latest @ . begin
        over over cell + @ 
        IMMEDIATE 1 - and
        = if swap drop TRUE
          else @ dup 
               if FALSE 
               else swap drop TRUE 
               then
          then 
        until 
        dup if cell + cell + then
        ;

 : ' hash seek ;

 : postpone ' , ;

 &&

 sp0 rp0 %S

&&

File: scrs/t1.fs

\ where is the hash

 : LINK>HASH CELL + ; 

 SEE 

\ where is the body

 : LINK>BODY CELL + CELL + ; 

 SEE 

 \ make a header

 : :NAME HERE : 0 STATE ! ; 

 SEE

 \ make a body

 : :NONAME HERE 1 STATE ! ; 

 SEE

 \ : HASH HERE :NAME SWAP HEAP ! CELL + @ ; 

 : DJB2 LIT [ 1024 DUP DUP + DUP + + 256 + 4 + 1 + , ] ; 

 SEE

 DJB2 . SPACE CR 

 BL . SPACE CR 

 \ BYE

 : HASH 
        BL BEGIN KEY OVER OVER = NOT UNTIL 
        DJB2 BEGIN 
                DUP DUP + DUP + DUP + DUP + DUP + + XOR >R
                KEY OVER OVER = 
                IF TRUE ELSE R> FALSE THEN
        UNTIL
        DROP DROP R>  
        ;
 
 SEE 

 : FIND LATEST @ BEGIN
        OVER OVER CELL + @
        ISNEGATIVE 1 - AND
        = IF SWAP DROP TRUE EXIT THEN
        @ DUP 0 
        = IF SWAP DROP FALSE EXIT THEN
        AGAIN ; 
 
 SEE 

 \ BYE

 HASH HASH . ." "

 HASH U@ . ." "

 HASH 0# . ." "

 : ' HASH FIND IF CELL + CELL + THEN ;  
 
 : '= ' ; IMMEDIATE  

 : POSTPONE ' , ; IMMEDIATE 

 \ crude pointer for CREATE DOES>

 : >BODY ['] LIT , HERE CELL + , 0 , ;

 \ from eforth, first EXIT is reserved for DOES> 

 : CREATE :NAME 
        ['] LIT , 
        HERE CELL + CELL + CELL + , 
        HERE >BODY ! 
        ['] EXIT , 
        ['] EXIT , 
        LATEST ! ; 
 
 : DOES> R> >BODY @ ! ; 

 : <BUILDS CREATE 0 , ; 

 : VARIABLE CREATE CELL ALLOT ; 

 : CONSTANT CREATE , DOES> @ ; 
 
 : BUFFER CREATE ALLOT ; 

 : ARRAY CREATE ALLOT DOES> + @ ; 

 : VALUE CREATE , DOES> @ ; 
 
 : TO ' CELL + @ 
        STATE @ 
        IF ['] LIT , , ['] ! , 
        ELSE ! THEN ; 


